pennylane/devices/default_mixed.py

```diff
@@ -118,13 +119,55 @@ def _create_basis_state(self, index):
         rho = self._asarray(rho, dtype=self.C_DTYPE)
         return self._reshape(rho, [2] * (2 * self.num_wires))
 
+    @classmethod
+    def capabilities(cls):
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            model="mixed",
+            returns_state=True,
+        )
+        return capabilities
+
     @property
     def state(self):
         """Returns the state density matrix of the circuit prior to measurement"""
         dim = 2 ** self.num_wires
         # User obtains state as a matrix
         return self._reshape(self._pre_rotated_state, (dim, dim))
 
+    def density_matrix(self, wires=None):
+        """Return the reduced density matrix representing a computational basis state
+         over the given wires.
+
+        Args:
+            wires (Wires): wires of the reduced system
+
+        Returns:
+            array[complex]: complex array of shape ``(2 ** len(wires), 2 ** len(wires))``
+            representing the reduced density matrix of the state prior to measurement.
+        """
+
+        # Return the full density matrix if all the wires are given
+        if wires == self.wires:
+            return self.state
+
+        wires = list(wires.labels)
+
+        traced_system = [x for x in list(self.wires.labels) if x not in wires]
+        traced_wires = Wires(traced_system)
+
+        # Trace first subsystem by applying kraus operators of the partial trace
+        tr_op = np.asarray(np.eye(2).reshape(1, 4), dtype=complex)
+        tr_op = self._reshape(tr_op, (2, 1, 2))
+
+        self._apply_channel(tr_op, traced_wires[0])
+
+        # Trace next subsystem by applying kraus operators of the partial trace
+        for traced_wire in traced_wires[1:]:
+            self._apply_channel(tr_op, traced_wire)
```

Comment:
This is a clever way of implementing it!

However, I wonder if there is a more efficient method that avoids a slow Python for loop. For example, by using `np.einsum` or `np.tensordot` to perform the partial trace of _all_ non-specified wires simultaneously.

Reply:
The number of Kraus operators for the partial trace is following 2^(#number all non-specified wires) and we need to construct all of them by using `np.tensordot`. I think it leads to a bigger number of operations overall compared to applying successively the Kraus operator for one system and the code will be more complicated to understand.